learning/tour-of-beam/frontend/lib/pages/tour/state.dart

```diff
@@ -0,0 +1,172 @@
+import 'package:get_it/get_it.dart';
+import 'package:playground_components/playground_components.dart';
+
+import '../../cache/unit_content.dart';
+import '../../models/unit.dart';
+import '../../models/unit_content.dart';
+import 'controllers/content_tree.dart';
+import 'path.dart';
+
+// Copied from Playground's config.g.dart
+// TODO(alexeyinkin): Read some config when backend's URL is emitted from
+//  an infrastructure workflow.
+
+const String kApiClientURL = 'https://router.34.160.22.119.nip.io';
+const String kApiJavaClientURL =
+    'https://backend-java-beta-dot-apache-beam-testing.appspot.com';
+const String kApiGoClientURL =
+    'https://backend-go-beta-dot-apache-beam-testing.appspot.com';
+const String kApiPythonClientURL =
+    'https://backend-python-beta-dot-apache-beam-testing.appspot.com';
+const String kApiScioClientURL =
+    'https://backend-scio-beta-dot-apache-beam-testing.appspot.com';
+
+class TourNotifier extends ChangeNotifier with PageStateMixin<void> {
+  final ContentTreeController contentTreeController;
+  final PlaygroundController playgroundController;
+  final _unitContentCache = GetIt.instance.get<UnitContentCache>();
+  UnitContentModel? _currentUnitContent;
+
+  TourNotifier({
+    required String initialSdkId,
+    List<String> initialTreeIds = const [],
+  })  : contentTreeController = ContentTreeController(
+          initialSdkId: initialSdkId,
+          initialTreeIds: initialTreeIds,
+        ),
+        playgroundController = _createPlaygroundController(initialSdkId) {
+    contentTreeController.addListener(_onContentTreeControllerChanged);
+    _unitContentCache.addListener(_onChanged);
+  }
+
+  @override
+  PagePath get path => TourPath(
+        sdkId: contentTreeController.sdkId,
+        treeIds: contentTreeController.treeIds,
+      );
+
+  void _onContentTreeControllerChanged() {
+    _onChanged();
+  }
+
+  void _onChanged() {
+    final currentNode = contentTreeController.currentNode;
+    if (currentNode is UnitModel) {
+      final content = _unitContentCache.getUnitContent(
+        contentTreeController.sdkId,
+        currentNode.id,
+      );
+
+      _setCurrentUnitContent(content);
+    } else {
+      _emptyPlayground();
+    }
+
+    notifyListeners();
+  }
+
+  UnitContentModel? get currentUnitContent => _currentUnitContent;
+
+  void _setCurrentUnitContent(UnitContentModel? content) {
+    if (content == _currentUnitContent) {
+      return;
+    }
+
+    _currentUnitContent = content;
+
+    if (content == null) {
+      return;
+    }
+
+    final taskSnippetId = content.taskSnippetId;
+    if (taskSnippetId == null) {
+      _emptyPlayground();
+      return;
+    }
+
+    playgroundController.examplesLoader.load(
+      ExamplesLoadingDescriptor(
+        descriptors: [
+          UserSharedExampleLoadingDescriptor(snippetId: taskSnippetId),
+        ],
+      ),
+    );
+  }
+
+  // TODO(alexeyinkin): Hide the entire right pane instead.
+  void _emptyPlayground() {
```

Comment:
Methods should be named using some action. Suppose `_loadEmptyExample` will be better

Reply:
'empty' is a verm: https://www.merriam-webster.com/dictionary/empty#dictionary-entry-2